src/config.rs

```diff
@@ -20,50 +18,62 @@ pub enum ToolSpec {
         gitlab: String,
         version: VersionReq,
     },
+    Artifactory {
+        artifactory: String,
+        path: String,
+        version: VersionReq,
+    },
 }
 
 impl ToolSpec {
     pub fn cache_key(&self) -> CiString {
         match self {
             ToolSpec::Github { github, .. } => CiString(github.clone()),
             ToolSpec::Gitlab { gitlab, .. } => CiString(format!("gitlab@{}", gitlab)),
+            ToolSpec::Artifactory { artifactory, .. } => CiString(artifactory.clone()),
         }
     }
 
-    pub fn source(&self) -> &str {
+    pub fn source(&self) -> String {
         match self {
             ToolSpec::Github { github: source, .. } | ToolSpec::Gitlab { gitlab: source, .. } => {
-                source
+                source.to_owned()
+            }
+            ToolSpec::Artifactory {
+                artifactory: source,
+                path,
+                ..
+            } => {
+                format!("{}/{}", source, path) //might want to use Url crate to enforce propper paths
             }
         }
     }
 
     pub fn version(&self) -> &VersionReq {
         match self {
-            ToolSpec::Github { version, .. } | ToolSpec::Gitlab { version, .. } => version,
+            ToolSpec::Github { version, .. }
+            | ToolSpec::Gitlab { version, .. }
+            | ToolSpec::Artifactory { version, .. } => version,
         }
     }
 
     pub fn provider(&self) -> Provider {
         match self {
             ToolSpec::Github { .. } => Provider::Github,
             ToolSpec::Gitlab { .. } => Provider::Gitlab,
+            ToolSpec::Artifactory { .. } => Provider::Artifactory,
         }
     }
 }
 
 impl fmt::Display for ToolSpec {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(
-            f,
-            "{}.com/{}@{}",
-            match self {
-                ToolSpec::Github { .. } => "github",
-                ToolSpec::Gitlab { .. } => "gitlab",
-            },
-            self.source(),
-            self.version(),
-        )
+        let res = match self {
+            ToolSpec::Github { .. } => format!("github.com/{}@{}", self.source(), self.version()),
+            ToolSpec::Gitlab { .. } => format!("gitlab.com/{}@{}", self.source(), self.version()),
+            ToolSpec::Artifactory { .. } => format!("{}@{}", self.source(), self.version()),
+        };
+        write!(f, "{}", res)
```

Comment:
Again, any rusty way to simplify it like before?

Reply:
I think you could inline the match like above, but this reads just fine to me